### DISCUSSION TITLE
Does the SAML reply URL need to include the registration ID?

### DISCUSSION THREAD

**Discussion by andrey-dubnik** [structurizr/onpremises#105] [Q&A] [answered]
### Description

The reply URL changed in the latest versions
From '/login/saml2/sso'
To '/login/saml2/sso/structurizr'

This is either a bug or some doc update may be required

### Steps to reproduce

Configure SAML

### Severity

Minor

### Priority

I have no budget and there's no rush, please fix this for free

**Comment by andrey-dubnik** [comment]
latest version which have SSO reply in line with documentation is 3263

**Comment by simonbrowndotje** [comment]
Correct, the SAML configuration changed in [2024.01.02 - see release notes](https://github.com/structurizr/onpremises/releases/tag/2024.01.02), and [https://docs.structurizr.com/onpremises/authentication/saml](https://docs.structurizr.com/onpremises/authentication/saml) has the new configuration instructions. Feel free to open a docs PR if you feel more is required.

**Comment by andrey-dubnik** [comment]
Thanks @simonbrowndotje, current version of the documentation states

> Register the Structurizr on-premises application with your Identity Provider. When doing this, you will need a “Reply URL”, which is of the form {structurizr.url}/login/saml2/sso

Where it actually expects `{structurizr.url}/login/saml2/sso/structurizr` now, at least once I have updated a version I have started to get the error message about the new reply URL from my IDP (AzureAD).

Hence I thought it is a bug which introduced extra to the SSO reply or a doc update. I'm happy to do a PR if you confirm the doc needs a change from `/sso` to `/sso/structurizr`

**Comment by andrey-dubnik** [comment]
Doc update would include following

Register the Structurizr on-premises application with your Identity Provider. When doing this, you will need a “Reply URL”, which is of the form {structurizr.url}/login/saml2/sso/{registrationId}

**Comment by simonbrowndotje** [comment]
So just to clarify:

- Build 3263 uses the old Spring SAML library, which required a reply URL of the form `{structurizr.url}/saml/SSO`.
- Build 2024.01.02 uses the new Spring SAML library, which requires a reply URL of the form `{structurizr.url}/login/saml2/sso`, but you're saying that this may be `{structurizr.url}/login/saml2/sso/{registrationId}` instead.

You may be correct, but this doesn't seem to be the case with any of the test apps I have on various identity providers; for example (this is Azure AD):

<img width="789" alt="Screenshot 2024-03-04 at 15 39 19" src="https://github.com/structurizr/onpremises/assets/1009874/89026be3-0da7-4a0e-97b1-7c172132fa6d">

If I set the reply URL to `{structurizr.url}/login/saml2/sso/example`, then the `structurizr.saml.registrationId` property does need to be set to `example` in the `structurizr.properties` file, otherwise I do see an Azure AD error. But setting the reply URL to `{structurizr.url}/login/saml2/sso` seems to work for me irrespective of what the registration ID is set to. I think this requires more investigation before updating the docs.

**Comment by andrey-dubnik** [comment]
Found the misconfiguration on my end (need to use glasses more often) - I had `/saml/SSO` setup on my SSO application which now needs a change to a different URL in line with the new requirements.

Although the change may have a positive impact in theory my IDP was suggesting that I need a ReplyURL set as `sso/structurizr` which makes sense as  accordingly to the [Spring documentation](https://docs.spring.io/spring-security/reference/servlet/saml2/login/authentication.html#saml2-response-processing-endpoint) the default endpoint is `/login/saml2/sso/{registrationId}` but it can be changed.

I guess either `/sso` or `/sso/{registrationId}` may be valid reply URLs but I'm yet to test that and it may take few days to come through the change requests.

**Comment by andrey-dubnik** [comment, marked as answer]
Tested with Reply URLs`/login/saml2/sso/{registrationId}` and `/login/saml2/sso` and both reply URLs are working. I guess documentation does not require an update.

**Comment by fiddlerpianist** [comment]
In our experience, we *had* to set the value of `registrationId` to something (i.e., the empty string caused the current version of Structurizr to crash), so we were unable to get `/login/saml2/sso` working as a valid reply URL.

**Comment by abalosache** [comment]
@simonbrowndotje other problem that I face now with on prem docker image is that azure ad doesn't allow you to set the replay url without https anymore 
![image](https://github.com/user-attachments/assets/2bb96086-2579-4026-90c6-2b1064665fe6)
 and on-prem image if you deploy it in a k8s cluster node it will try to send a replay that is under http doesn t matter if you do a redirect to https from nginx or from other layer 

![image](https://github.com/user-attachments/assets/19f947dd-2615-48a1-9790-5d11a2035c9a)
